src/core/symbology/qgssymbol.cpp

```diff
@@ -116,57 +127,82 @@ QPolygonF QgsSymbol::_getLineString( QgsRenderContext &context, const QgsCurve &
   }
 
   // remove non-finite points, e.g. infinite or NaN points caused by reprojecting errors
-  pts.erase( std::remove_if( pts.begin(), pts.end(),
-                             []( const QPointF point )
+  pts.filterVertices( []( const QgsPoint point )
   {
-    return !std::isfinite( point.x() ) || !std::isfinite( point.y() );
-  } ), pts.end() );
+    if ( point.is3D() )
+    {
+      return std::isfinite( point.x() ) && std::isfinite( point.y() ) && std::isfinite( point.z() );
+    }
+    else
+    {
+      return std::isfinite( point.x() ) && std::isfinite( point.y() );
+    }
+  } );
 
   if ( clipToExtent && nPoints > 1 && context.flags() & QgsRenderContext::ApplyClipAfterReprojection )
   {
     // early clipping was not possible, so we have to apply it here after transformation
     const QgsRectangle e = context.mapExtent();
     const double cw = e.width() / 10;
     const double ch = e.height() / 10;
-    const QgsRectangle clipRect( e.xMinimum() - cw, e.yMinimum() - ch, e.xMaximum() + cw, e.yMaximum() + ch );
-    pts = QgsClipper::clippedLine( pts, clipRect );
+    const QgsBox3d clipRect( QgsPoint( e.xMinimum() - cw, e.yMinimum() - ch ), QgsPoint( e.xMaximum() + cw, e.yMaximum() + ch ) ); // TODO also need to be clipped according to z axis
+    pts = QgsClipper::clipped3dLine( pts, clipRect );
   }
 
-  QPointF *ptr = pts.data();
-  for ( int i = 0; i < pts.size(); ++i, ++ptr )
+  QPolygonF out = pts.asQPolygonF();
+  QPointF *ptr = out.data();
+  for ( int i = 0; i < out.size(); ++i, ++ptr )
   {
     mtp.transformInPlace( ptr->rx(), ptr->ry() );
   }
 
-  return pts;
+  printf( "_getLineString(out): %d\n", out.size() );
```

Comment:
printf() debug statements to remove here and at other locations in that file

Reply:
Thanks, yet I was sure to have removed them! :)